### PR TITLE
Add aten mkldnn view operator

### PR DESCRIPTION
**Pull request by bddppq**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #19633 Add is_mkldnn to at::Tensor&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D15053320/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #19204 Add aten mkldnn conv2d operator&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D14857513/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #19205 Add aten mkldnn ops: relu, max_pool2d and avg_pool2d&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D14850598/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #19206 Add aten mkldnn batch_norm operator&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D14887205/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #19207 Add aten mkldnn add operator&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D14889477/)
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#19209 Add aten mkldnn view operator**&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D14894545/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #19210 Add aten mkldnn linear operator&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D14901641/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #19648 Adjust resnext run script&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D14889481/)

Pull Request resolved: https://github.com/pytorch/pytorch/pull/19209

Differential Revision: [D14894545](https://our.internmc.facebook.com/intern/diff/D14894545/)

### CODE REVIEW

File: test/test_mkldnn.py [hunk 1]
```diff
@@ -182,6 +182,15 @@ def test_add(self):
         mx += my
         self.assertEqual(x, mx.to_dense())
 
+    def test_view(self):
+        x = torch.randn(3, 4, 5, dtype=torch.float32) * 10
+        size = (x.size(0), -1)
+
+        self.assertEqual(
+            x.view(size),
```
Comment (gchanan) [inline test/test_mkldnn.py]: this doesn't actually check that view shares memory.

File: aten/src/ATen/native/mkldnn/View.cpp [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+namespace at {
+namespace native {
+
+Tensor mkldnn_view(const Tensor& self, IntArrayRef size) {
+  auto inferred_size = at::infer_size(size, self.numel());
+  const ideep::tensor& x = itensor_from_mkldnn(self);
+  ideep::tensor y;
+  ideep::direct_copy::compute<AllocForMKLDNN>(x, y);
+  y.reshape({inferred_size.cbegin(), inferred_size.cend()});
```
Comment (gchanan) [inline aten/src/ATen/native/mkldnn/View.cpp]: this looks like a copy, not a view.

**Review by gchanan — CHANGES_REQUESTED** [review]

File: aten/src/ATen/native/mkldnn/View.cpp (hunk 2 above)
Comment (bddppq) [reply aten/src/ATen/native/mkldnn/View.cpp]: yes it doesn't, but for opaque tensor I thought it's expected?

File: test/test_mkldnn.py (hunk 1 above)
Comment (jgong5) [reply test/test_mkldnn.py]: Is memory sharing an API semantic of `view` (and others like `narrow`) or just a zero-copy optimization?

Comment (dzhulgakov) [reply test/test_mkldnn.py]: So in general, it's an API semantics of .view as people rely on taking view and then applying mutability: https://pytorch.org/docs/stable/tensors.html#torch.Tensor.view

Would it be plausible to NOT implement view and instead implement `reshape` only?

(we should implement `reshape` anyway - it allows to return a copy)

File: aten/src/ATen/native/mkldnn/TensorShape.cpp [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+} // namespace at
+
+#else // AT_MKLDNN_EBABLED
+
+#include <ATen/native/mkldnn/MKLDNNCommon.h>
+
+namespace at {
+namespace native {
+
+Tensor mkldnn_view(const Tensor& self, IntArrayRef size) {
```
Comment (dzhulgakov) [inline aten/src/ATen/native/mkldnn/TensorShape.cpp]: nit: I guess you can move it out of the #ifdef as the error is regardless of whether mkldnn is compiled in

File: aten/src/ATen/native/native_functions.yaml [hunk 4]
```diff
@@ -1517,6 +1517,12 @@
   variants: function, method
   device_guard: False
 
+- func: mkldnn_reshape(Tensor self, int[] shape) -> Tensor
+  device_guard: False
+  requires_tensor: True
```
Comment (dzhulgakov) [inline aten/src/ATen/native/native_functions.yaml]: you don't need `requires_tensor` if you have `dispatch`

**Review by dzhulgakov — APPROVED** [review]

File: aten/src/ATen/native/native_functions.yaml (hunk 4 above)
Comment (bddppq) [reply aten/src/ATen/native/native_functions.yaml]: yep I know this, just for documentation purpose :-)

File: aten/src/ATen/native/mkldnn/TensorShape.cpp (hunk 3 above)
Comment (bddppq) [reply aten/src/ATen/native/mkldnn/TensorShape.cpp]: The error messages are different though so I would prefer to keep two these two places.

**Pull request closed without merging**